### ISSUE TITLE
Invalid URI throws exception when NO_PROXY is set in environment

### ISSUE THREAD

**Issue by jondistad** [request/request#1595]
```
> request.post({uri: 'invalid'}, function(err, res){ })
/.../node_modules/request/lib/getProxyFromURI.js:5
  return hostname.replace(/^\.*/, '.').toLowerCase()
                 ^
TypeError: Cannot read property 'replace' of null
  at formatHostname (/.../node_modules/request/lib/getProxyFromURI.js:5:18)
  at uriInNoProxy (/.../node_modules/request/lib/getProxyFromURI.js:21:18)
  at getProxyFromURI (/.../node_modules/request/lib/getProxyFromURI.js:55:25)
  at Request.init (/.../node_modules/request/request.js:447:18)
```

The URI validation happens on [line 457 of request.js](https://github.com/request/request/blob/master/request.js#L457), but when `NO_PROXY` is set in the environment, the hostname is parsed out of the URI before the validation happens. If an invalid URI has no hostname, then the above null error is thrown, which is no good for asynchronous code.

It looks like it'd be possible to move the URI validation above the proxy and tunnel bits, so that the error can be emitted normally.

**Comment by simov** [comment]
Good call @jondistad fixed in #1595

**Comment by jondistad** [comment]
Excellent, thanks!
